### DISCUSSION TITLE
View-Model-Controller structure

### DISCUSSION THREAD

**Discussion by xsebek** [swarm-game/swarm#2091] [General]
I wanted to add a new feature to the UI state, but I found that it had become a bit messy.

It seems we have four approaches to the UI code structure at the same time:

* The initial approach of single Model, View, and Controller files (now 350, 1600, and 800 lines long)
* The split into smaller files in Model, View, and Controller directories
  * This was not enough, so for example I have made a subdirectory EventHandlers in Controller directory
* This does not group code by purpose, so there is Inventory, Launch and editor directory along with a few toplevel files

How do we want to structure code going forward?

**Comment by xsebek** [comment]
My personal favourites are the subdirectories for MVC and nesting by purpose _together_.

I am not entirely sure which one should be top-level, I guess it might be preferable to have MVC top in Haskell.

**Reply by byorgey** [reply]
Yeah, in retrospect I'm not sure the standard MVC architecture is all that relevant in Haskell.  I guess it makes the most sense when the model lives in a database on one server, the controller is code running on another server, and so on.  But just for organizing code I think it probably makes more sense to instead group code by purpose, as you say.
